vault/resource_ad_secret_backend.go

```diff
@@ -0,0 +1,645 @@
+		data["deny_null_bind"] = v
+	}
+	if v, ok := d.GetOkExists("discoverdn"); ok {
+		data["discoverdn"] = v
+	}
+	if v, ok := d.GetOkExists("formatter"); ok {
+		data["formatter"] = v
+	}
+	if v, ok := d.GetOkExists("groupattr"); ok {
+		data["groupattr"] = v
+	}
+	if v, ok := d.GetOkExists("groupdn"); ok {
+		data["groupdn"] = v
+	}
+	if v, ok := d.GetOkExists("groupfilter"); ok {
+		data["groupfilter"] = v
+	}
+	if v, ok := d.GetOkExists("insecure_tls"); ok {
+		data["insecure_tls"] = v
+	}
+	if v, ok := d.GetOkExists("last_rotation_tolerance"); ok {
+		data["last_rotation_tolerance"] = v
+	}
+	if v, ok := d.GetOkExists("length"); ok {
+		data["length"] = v
+	}
+	if v, ok := d.GetOkExists("max_ttl"); ok {
+		data["max_ttl"] = v
+	}
+	if v, ok := d.GetOkExists("password_policy"); ok {
+		data["password_policy"] = v
+	}
+	if v, ok := d.GetOkExists("request_timeout"); ok {
+		data["request_timeout"] = v
+	}
+	if v, ok := d.GetOkExists("starttls"); ok {
+		data["starttls"] = v
+	}
+	if v, ok := d.GetOkExists("tls_max_version"); ok {
+		data["tls_max_version"] = v
+	}
+	if v, ok := d.GetOkExists("tls_min_version"); ok {
+		data["tls_min_version"] = v
+	}
+	if v, ok := d.GetOkExists("ttl"); ok {
+		data["ttl"] = v
+	}
+	if v, ok := d.GetOkExists("upndomain"); ok {
+		data["upndomain"] = v
+	}
+	if v, ok := d.GetOkExists("url"); ok {
+		data["url"] = v
+	}
+	if v, ok := d.GetOkExists("use_pre111_group_cn_behavior"); ok {
+		data["use_pre111_group_cn_behavior"] = v
+	}
+	if v, ok := d.GetOkExists("use_token_groups"); ok {
+		data["use_token_groups"] = v
+	}
+	if v, ok := d.GetOkExists("userattr"); ok {
+		data["userattr"] = v
+	}
+	if v, ok := d.GetOkExists("userdn"); ok {
+		data["userdn"] = v
+	}
+
+	configPath := fmt.Sprintf("%s/config", backend)
+	log.Printf("[DEBUG] Writing %q", configPath)
+	if _, err := client.Logical().Write(configPath, data); err != nil {
+		return fmt.Errorf("error writing %q: %s", configPath, err)
+	}
+	log.Printf("[DEBUG] Wrote %q", configPath)
+	return readConfigResource(d, meta)
+}
+
+func readConfigResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	path := d.Id()
+	log.Printf("[DEBUG] Reading %q", path)
+
+	mountResp, err := client.Sys().MountConfig(path)
+	if err != nil {
+		return fmt.Errorf("error reading %q: %s", path, err)
+	}
+
+	d.Set("backend", d.Id())
+
+	d.Set("default_lease_ttl_seconds", mountResp.DefaultLeaseTTL)
+	d.Set("max_lease_ttl_seconds", mountResp.MaxLeaseTTL)
+
+	configPath := fmt.Sprintf("%s/config", d.Id())
+	log.Printf("[DEBUG] Reading %q", configPath)
+
+	resp, err := client.Logical().Read(configPath)
+	if err != nil {
+		return fmt.Errorf("error reading %q: %s", configPath, err)
+	}
+	log.Printf("[DEBUG] Read %q", configPath)
+	if resp == nil {
+		log.Printf("[WARN] %q not found, removing from state", configPath)
+		d.SetId("")
+		return nil
+	}
+
+	if val, ok := resp.Data["anonymous_group_search"]; ok {
```

Comment:
Same question on using a for loop here. I see there are some slightly more specific error messages, but I don't think they add much value compared to being confident that the code is doing the right thing.

Reply:
^ I think this would make a good cleanup across the board. Something we could target in the future.

Reply:
They don't always contain data and reading `ad/config` doesn't return them if they're null. Here's a read on the `ad/config` the bare minimum settings:

```
vault read ad/config
Key                             Value
---                             -----
binddn                          CN=Administrator,CN=Users,DC=corp,DC=example,DC=net
certificate                     n/a
formatter                       n/a
insecure_tls                    true
last_rotation_tolerance         5
length                          64
max_ttl                         768h
policy_name                     n/a
starttls                        false
tls_max_version                 tls12
tls_min_version                 tls12
ttl                             768h
upndomain                       n/a
url                             ldaps://ad
use_pre111_group_cn_behavior    false
userdn                          CN=Users,DC=corp,DC=example,DC=net
```